### PR TITLE
Big messages effectiveness

### PR DESCRIPTION
**Pull request by baktrius**
Updates regarding specification of effectiveness for group communication functions in case of processing big messages.
Improve measuring of effectiveness (delay is now applied only to one function).

### CODE REVIEW

File: examples/bare_reduce.c [hunk 1]
```diff
@@ -17,9 +19,14 @@ int main(int argc, char **argv)
     MIMPI_Init(false);
     int const world_rank = MIMPI_World_rank();
 
+    const char *delay = getenv("DELAY");
+    if (delay)
+        setenv(WRITE_VAR, delay, true);
+
```
Comment (wprzytula) [inline examples/bare_reduce.c:25]: Calls to `setenv`, `getenv`, `unsetenv` should be error-checked.

File: tests/effectiveness/barrier_15_delay_100ms.self [hunk 2]
```diff
@@ -1,4 +1,4 @@
-MIMPI_WRITE_DELAY=100 timeout 2s ./mimpirun 15 examples_build/bare_barrier
+DELAY=100 timeout 2s ./mimpirun 15 examples_build/bare_barrier
```
Comment (wprzytula) [inline tests/effectiveness/barrier_15_delay_100ms.self:1]: Why not set this as `CHANNELS_WRITE_DELAY` and avoid subsequent fiddling with `setenv` in examples?

**Review by wprzytula — CHANGES_REQUESTED** [review]

File: tests/effectiveness/barrier_15_delay_100ms.self (hunk 2 above)
Comment (baktrius) [reply tests/effectiveness/barrier_15_delay_100ms.self:1]: The way it is done now, DELAY is set only for duration of function which execution time we want to measure. Therefore one can send some helper messages in MIMPI_Init or MIMPI_Finalise if one wants.

File: examples/bare_reduce.c (hunk 1 above)
Comment (baktrius) [reply examples/bare_reduce.c]: Finally I used simple asserts to check for system calls success to not copy whole ASSERT_SYS_OK functionality.

File: examples/bare_barrier.c [hunk 3]
```diff
@@ -13,9 +14,9 @@ int main(int argc, char **argv)
     fflush(stdout);
     const char *delay = getenv("DELAY");
     if (delay)
-        setenv(WRITE_VAR, delay, true);
+        assert(setenv(WRITE_VAR, delay, true) == 0);
```
Comment (wprzytula) [inline examples/bare_barrier.c:17]: Relying on `assert`s here has one drawback: compilation with `-DNDEBUG` will make those statements no-ops. Of course we don't use this flag, but this is why asserts should be rather called on return values, not on the statements themselves. We can afford not checking a return code, but not skipping the whole statement.

Comment (baktrius) [reply examples/bare_barrier.c:17]: Sure, I feel a little ashamed now. Using asserts like that is probably first position on every c noob's errors list in existence. Thanks for pointing it out. I will quickly fix it before someone else out there will see it.

File: examples/bare_reduce.c [hunk 4]
```diff
@@ -1,20 +1,42 @@
...
+        assert(recv_data);
+        ASSERT_MIMPI_OK(MIMPI_Reduce(data, recv_data, data_size, MIMPI_SUM, 0));
+        for (int i = 1; i < data_size; ++i)
+            assert(recv_data[i] == recv_data[0]);
+        printf("Number: %d\n", recv_data[0]);
         fflush(stdout);
+        free(recv_data);
     }
+    else
+        ASSERT_MIMPI_OK(MIMPI_Reduce(data, NULL, data_size, MIMPI_SUM, 0));
```
[the remark this answers is not in the document]
Comment (baktrius) [reply examples/bare_reduce.c]: Done

**Review by wprzytula — APPROVED** [review]

**Pull request merged**